### ISSUE TITLE
ixgbe sfp+ always gets "Link Down" error

### ISSUE THREAD

**Issue by shunghsiyu** [cisco-system-traffic-generator/trex-core#527]
This may possibly be a duplicate of #511, however I decided to open a new issue rather than reopen since:
1. On my system is always ended up with an error, rather than occasionally
2. Testing with `testpmd` that comes with several different version of DPDK (v18.11, v19.11, v20.05) shows that DPDK can correctly bring the link up, i.e. this might be a TRex problem rather than a DPDK one

Following the advise here https://github.com/cisco-system-traffic-generator/trex-core/issues/511#issuecomment-662882009 of **downgrading to TRex v2.79 works**, but it would be even better to have the latest version of TRex work on my system as well.

---

My setup consists of two workstations connected directly with optic fiber, the NICs on both of the workstations is X520-2 (Intel Corporation 82599ES 10-Gigabit SFI/SFP+ Network Connection [8086:10fb]). I'm not sure what SFP+ module is used on both of them. On the server side the workstation is running Nginx, and the NIC is controlled by the kernel with the `ixgbe` driver; on the client side the workstation is running TRex with the net_ixgbe driver (through vfio-pci).

```
$  sudo lspci -s 0000:04:00.1 -vnn # Showing just the output of on workstation since both is the same
0000:04:00.1 Ethernet controller [0200]: Intel Corporation 82599ES 10-Gigabit SFI/SFP+ Network Connection [8086:10fb] (rev 01)
        Subsystem: Intel Corporation Ethernet Server Adapter X520-2 [8086:000c]
        Flags: fast devsel, IRQ 17, NUMA node 0, IOMMU group 24
        Memory at 38003fd00000 (64-bit, prefetchable) [disabled] [size=512K]
        I/O ports at 2000 [disabled] [size=32]
        Memory at 38003fe00000 (64-bit, prefetchable) [disabled] [size=16K]
        Capabilities: [40] Power Management version 3
        Capabilities: [50] MSI: Enable- Count=1/1 Maskable+ 64bit+
        Capabilities: [70] MSI-X: Enable- Count=64 Masked-
        Capabilities: [a0] Express Endpoint, MSI 00
        Capabilities: [100] Advanced Error Reporting
        Capabilities: [140] Device Serial Number 00-1b-21-ff-ff-3c-9e-2c
        Capabilities: [150] Alternative Routing-ID Interpretation (ARI)
        Capabilities: [160] Single Root I/O Virtualization (SR-IOV)
        Kernel driver in use: vfio-pci
        Kernel modules: ixgbe
$ sudo ethtool ixgbe1
Settings for ixgbe1:
        Supported ports: [ FIBRE ]
        Supported link modes:   1000baseT/Full
                                10000baseT/Full
        Supported pause frame use: Symmetric 
        Supports auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Speed: 10000Mb/s
        Duplex: Full
        Auto-negotiation: on
        Port: FIBRE
        PHYAD: 0
        Transceiver: internal
        Supports Wake-on: d
        Wake-on: d
        Current message level: 0x00000007 (7)
                               drv probe link
        Link detected: yes
```

TRex configuration is as follow

```yaml
- port_limit      : 2
  version         : 2
  low_end: true  # not sure if needed
  interfaces    : ["0000:04:00.1","dummy"] 
  port_info       :  # MAC redacted
          - src_mac: xx:xx:xx:xx:xx:xx
            dest_mac: xx:xx:xx:xx:xx:xx
          - src_mac: xx:xx:xx:xx:xx:xx
            dest_mac: xx:xx:xx:xx:xx:xx
```

And the following is the output I get when running [DNS basic example](https://trex-tgn.cisco.com/trex/doc/trex_manual.html#_dns_basic_example)

```
$ ./t-rex-64 -f cap2/dns.yaml -d 10 -v 7
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
Using configuration file /etc/trex_cfg.yaml 
 port limit     :  2 
 port_bandwidth_gb    :  10 
 if_mask        : None 
 is low-end : 1 
 low-end core : 0
 stack type :  
 thread_per_dual_if      : 1 
 if        :  0000:04:00.1, dummy,
 enable_zmq_pub :  1 
 zmq_pub_port   :  4500 
 m_zmq_rpc_port    :  4501 
 src     : 00:1b:21:3c:9e:2c
 dest    : 00:1b:21:3c:9e:31
 src     : 00:1b:21:3c:9e:31
 dest    : 00:1b:21:3c:9e:2c
 memory per 2x10G ports  
 MBUF_64                                   : 16380 
 MBUF_128                                  : 8190 
 MBUF_256                                  : 8190 
 MBUF_512                                  : 8190 
 MBUF_1024                                 : 8190 
 MBUF_2048                                 : 4095 
 MBUF_4096                                 : 128 
 MBUF_9K                                   : 512 
 TRAFFIC_MBUF_64                           : 65520 
 TRAFFIC_MBUF_128                          : 32760 
 TRAFFIC_MBUF_256                          : 8190 
 TRAFFIC_MBUF_512                          : 8190 
 TRAFFIC_MBUF_1024                         : 8190 
 TRAFFIC_MBUF_2048                         : 8928 
 TRAFFIC_MBUF_4096                         : 128 
 TRAFFIC_MBUF_9K                           : 512 
 MBUF_DP_FLOWS                             : 65536 
 MBUF_GLOBAL_FLOWS                         : 5120 
 no platform info 
CTimerWheelYamlInfo does not exist  
 flags           : 8010f00
 write_file      : 0
 verbose         : 7
 realtime        : 1
 flip            : 0
 cores           : 1
 single core     : 0
 flow-flip       : 0
 no clean close  : 0
 zmq_publish     : 1
 vlan mode       : 0
 client_cfg      : 0
 mbuf_cache_disable  : 0
 cfg file        : cap2/dns.yaml 
 mac file        :  
 out file        :  
 client cfg file :  
 duration        : 10 
 factor          : 1 
 mbuf_factor     : 1 
 latency         : 0 pkt/sec 
 zmq_port        : 4500 
 telnet_port     : 4501 
 expected_ports  : 2 
 tw_bucket_usec  : 20.000000 usec 
 tw_buckets      : 1024 usec 
 tw_levels       : 3 usec 
 port : 0 dst:00:1b:21:3c:9e:31  src:00:1b:21:3c:9e:2c
 port : 1 dst:00:1b:21:3c:9e:2c  src:00:1b:21:3c:9e:31
 port : 2 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 3 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 4 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 5 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 6 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 7 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 8 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 9 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 10 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 11 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 12 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 13 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 14 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 15 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 16 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 17 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 18 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 19 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 20 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 21 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 22 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 port : 23 dst:00:00:00:01:00:00  src:00:00:00:00:00:00
 Total Memory : 
 MBUF_64                                   : 81900 
 MBUF_128                                  : 40950 
 MBUF_256                                  : 16380 
 MBUF_512                                  : 16380 
 MBUF_1024                                 : 16380 
 MBUF_2048                                 : 13023 
 MBUF_4096                                 : 1024 
 MBUF_DP_FLOWS                             : 65536 
 MBUF_GLOBAL_FLOWS                         : 5120 
 get_each_core_dp_flows                    : 65536 
 Total memory                              :     113.49 Mbytes  
 there is no configuration file given 
DPDK args 
 xx  EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Static memory layout is selected, amount of reserved memory can be adjusted with -m or --socket-mem
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'VA'
EAL: No available hugepages reported in hugepages-1048576kB
--lcores  (0,1,2)@0  -n  4  --log-level  8  --master-lcore  0  -w  0000:04:00.1  --legacy-mem  
 EAL: Probing VFIO support...
EAL: VFIO support initialized
EAL: PCI device 0000:04:00.1 on NUMA socket 0
EAL:   probe driver: 8086:10fb net_ixgbe
EAL:   using IOMMU type 1 (Type 1)
EAL: Ignore mapping IO port bar(2)
                input : [0000:04:00.1, dummy]
                 dpdk : [0000:04:00.1]
             pci_scan : [0000:04:00.1]
                  map : [ 0, 255]
 TRex port mapping 
 ----------------- 
 TRex vport: 0 dpdk_rte_eth: 0 
 TRex vport: 1 dpdk_rte_eth: 255 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to DROP_QUE_FILTER 
 DPDK devices 1 : 1 
-----
 0 : vdev 0000:04:00.1 
-----
 Number of ports found: 2 (dummy among them: 1)


if_index : 0 
driver name : net_ixgbe 
min_rx_bufsize : 1024 
max_rx_pktlen  : 15872 
max_rx_queues  : 128 
max_tx_queues  : 64 
max_mac_addrs  : 127 
rx_offload_capa : 0x92e9f 
tx_offload_capa : 0xa03f 
rss reta_size   : 128 
flow_type_rss   : 0x38d34 
tx_desc_max     : 4096 
tx_desc_min     : 32 
rx_desc_max     : 4096 
rx_desc_min     : 32 
zmq publisher at: tcp://*:4500
 rx_data_q_num : 1 
 rx_drop_q_num : 1 
 rx_dp_q_num   : 0 
 rx_que_total : 2 
 --   
 rx_desc_num_data_q   : 4096 
 rx_desc_num_drop_q   : 64 
 rx_desc_num_dp_q     : 0 
 total_desc           : 4160 
 --   
 tx_desc_num     : 1024 
port 0 desc: 82599ES 10-Gigabit SFI/SFP+ Network Connection
 rx_qid: 0 (64) 
 rx_qid: 1 (4096) 
 wait 2 sec ..
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
EAL: Error - exiting with code: 1
  Cause:  One of the links is down 
```

Meanwhile dmesg on the server (using `ixgbe`) shows

```
[529233.599828] ixgbe 0000:04:00.1 ixgbe1: NIC Link is Up 10 Gbps, Flow Control: RX/TX
[529233.615929] ixgbe 0000:04:00.1 ixgbe1: NIC Link is Down
[529234.201100] ixgbe 0000:04:00.1 ixgbe1: NIC Link is Up 10 Gbps, Flow Control: RX/TX
[529236.291524] ixgbe 0000:04:00.1 ixgbe1: NIC Link is Down
```

**Comment by hhaim** [comment]
@shunghsiyu thanks for reopening the issue. We don't have a faulty hardware so we can't test it.

Could you check the folowing:

1) add ```--software``` to invocation 
2) add ```--no-flow-control-change``` to invocation 
3) with STL and ASTF modes 

The driver is the same as DPDK and with ```--software``` it is the standard configuration

**Comment by hhaim** [comment]
This is for STL and ASTF 
```./t-rex-64 -i -c 1 --software --iom 0 ```
```./t-rex-64 -i -c 1 --software --astf --iom 0 ```

**Comment by shunghsiyu** [comment]
@hhaim thanks for the speedy reply!
Since I only have SSH access it looks like I have to add `--no-scapy-server`.

In all of these commands it looks like the link is still down.
Do let me know if there's other test I can try or if I'm not testing correctly.

### STL

```
$ sudo ./t-rex-64 -i -c 1 --no-scapy-server --software --iom 0
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to ONE_QUE 
 Number of ports found: 2 (dummy among them: 1)
zmq publisher at: tcp://*:4500
 wait 2 sec ..
 WARNING : there is no link on one of the ports, interactive mode can continue
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
 number of ports         : 2 
 max cores for 2 ports   : 1 
 tx queues per port      : 3 
 -------------------------------
RX core uses TX queue number 65535 on all ports
 core, c-port, c-queue, s-port, s-queue, lat-queue
 ------------------------------------------
 1        0      0       1       0      0  
 -------------------------------
 *** TRex is shutting down - cause: 'CTRL + C detected'
 All cores stopped !! 
```

### ASTF

```
$ sudo ./t-rex-64 -i -c 1 --no-scapy-server --software --astf --iom 0
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to ONE_QUE 
 Number of ports found: 2 (dummy among them: 1)
zmq publisher at: tcp://*:4500
 wait 2 sec ..
 WARNING : there is no link on one of the ports, interactive mode can continue
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
 number of ports         : 2 
 max cores for 2 ports   : 1 
 tx queues per port      : 3 
 -------------------------------
RX core uses TX queue number 65535 on all ports
 core, c-port, c-queue, s-port, s-queue, lat-queue
 ------------------------------------------
 1        0      0       1       0      0  
 -------------------------------
 *** TRex is shutting down - cause: 'CTRL + C detected'
 All cores stopped !! 
```

### DNS basic example

#### Both

```
$ sudo ./t-rex-64 -f cap2/dns.yaml -d 10 --no-flow-control-change -c 1 --software --iom 0
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to ONE_QUE 
 Number of ports found: 2 (dummy among them: 1)
zmq publisher at: tcp://*:4500
 wait 2 sec ..
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
EAL: Error - exiting with code: 1
  Cause:  One of the links is down 
```

#### `--no-flow-control-change`

```
$ sudo ./t-rex-64 -f cap2/dns.yaml -d 10 --no-flow-control-change
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to DROP_QUE_FILTER 
 Number of ports found: 2 (dummy among them: 1)
zmq publisher at: tcp://*:4500
 wait 2 sec ..
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
EAL: Error - exiting with code: 1
  Cause:  One of the links is down 
```

#### --software

```
sudo ./t-rex-64 -f cap2/dns.yaml -d 10 -c 1 --software --iom 0
The ports are bound/configured.
Starting  TRex v2.83 please wait  ... 
 set driver name net_ixgbe 
 driver capability  : TCP_UDP_OFFLOAD  TSO 
 set dpdk queues mode to ONE_QUE 
 Number of ports found: 2 (dummy among them: 1)
zmq publisher at: tcp://*:4500
 wait 2 sec ..
port : 0 
------------
link         :  Link Down
promiscuous  : 0 
EAL: Error - exiting with code: 1
  Cause:  One of the links is down

**Comment by shunghsiyu** [comment]
For what its worth, I notice that when running `testpmd` (which does brings the link up), the "Port 0: link state change event" line only shows up about 3~5 seconds after the `testpmd>` prompt shows up.

Digging into DPDK code for testpmd, the "Port 0: link state change event" line seems to be outputted by the [`eth_event_callback`](https://github.com/DPDK/dpdk/blob/v19.11/app/test-pmd/testpmd.c#L2822-L2823), which is [registered](https://github.com/DPDK/dpdk/blob/v19.11/app/test-pmd/testpmd.c#L2853-L2856) as a [link status interrupt handler](https://doc.dpdk.org/guides-19.11/sample_app_ug/link_status_intr.html#interrupt-callback-registration). Looking around TRex's source code it seems the no interrupt handler is being set for ixgbe, so I'm wonder if processing link status interrupt is required for my setup to work.

But to be honest I cannot yet explain how processing link status interrupt in `testpmd` bring the link up. My guess is that the link status interrupt leads to [`setup_attached_port()`](https://github.com/DPDK/dpdk/blob/v19.11/app/test-pmd/testpmd.c#L2527)->[`reconfig()`](https://github.com/DPDK/dpdk/blob/v19.11/app/test-pmd/testpmd.c#L1319)->[`init_port_config()`](https://github.com/DPDK/dpdk/blob/v19.11/app/test-pmd/testpmd.c#L3053), and `init_port_config()` then brings the link up.

```
$ sudo /usr/bin/testpmd -w '0000:04:00.1' -- -i
EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'VA'
EAL: No available hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
EAL: PCI device 0000:04:00.1 on NUMA socket 0
EAL:   probe driver: 8086:10fb net_ixgbe
EAL:   using IOMMU type 1 (Type 1)
EAL: Ignore mapping IO port bar(2)
Interactive-mode selected
testpmd: create a new mbuf pool <mbuf_pool_socket_0>: n=203456, size=2176, socket=0
testpmd: preferred mempool ops selected: ring_mp_mc

Warning! port-topology=paired and odd forward ports number, the last port will pair with itself.

Configuring Port 0 (socket 0)
Port 0: 00:1B:21:3C:9E:2D
Checking link statuses...
testpmd> Done
# 3~5 seconds passes by before the next line shows up
Port 0: link state change event

testpmd> 
```

**Comment by hhaim** [comment]
@shunghsiyu trex is not using the link callback event. . Maybe there is a hidden bug in the driver that is solved due to the interrupt ..

**Comment by hhaim** [comment]
@shunghsiyu could you check L2forward which is simpler?

**Comment by shunghsiyu** [comment]
@hhaim l2fwd (DPDK v20.05) seems to work fine

When running l2fwd on workstation `david`, which is connect to `benjamin` and an unkown workstation that I don't control with an IP of 192.168.3.1, I can ping the unknown workstation on `benjamin`. (Please see below for the setup)

Here's how I ran l2fwd on `david`

```
shunghsiyu@david:/opt/dpdk/dpdk-20.05/build/examples> sudo LD_LIBRARY_PATH="/opt/dpdk/v20.05/lib64/" /opt/dpdk/dpdk-20.05/build/examples/dpdk-l2fwd -w '0000:04:00.0' -w '0000:04:00.1' -- -p 0x3 --no-mac-updating
EAL: Detected 8 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'VA'
EAL: No available hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
EAL:   using IOMMU type 1 (Type 1)
EAL: Ignore mapping IO port bar(1)
EAL: Ignore mapping IO port bar(2)
EAL: Ignore mapping IO port bar(5)
EAL: Probe PCI driver: net_ixgbe (8086:10fb) device: 0000:04:00.0 (socket 0)
EAL: Ignore mapping IO port bar(1)
EAL: Ignore mapping IO port bar(2)
EAL: Ignore mapping IO port bar(5)
EAL: Probe PCI driver: net_ixgbe (8086:10fb) device: 0000:04:00.1 (socket 0)
EAL: No legacy callbacks, legacy socket not created
MAC updating disabled
Lcore 0: RX port 0
Lcore 1: RX port 1
Initializing port 0... done:
Port 0, MAC address: XX:XX:XX:XX:XX:30

Initializing port 1... done:
Port 1, MAC address: XX:XX:XX:XX:XX:31

Checking link status.done
Port0 Link Up. Speed 10000 Mbps - full-duplex
Port1 Link Up. Speed 10000 Mbps - full-duplex
L2FWD: entering main loop on lcore 1
L2FWD:  -- lcoreid=1 portid=1
L2FWD: lcore 2 has nothing to do
L2FWD: lcore 3 has nothing to do
L2FWD: lcore 4 has nothing to do
L2FWD: lcore 6 has nothing to do
L2FWD: lcore 5 has nothing to do
L2FWD: lcore 7 has nothing to do
L2FWD: entering main loop on lcore 0
L2FWD:  -- lcoreid=0 portid=0

Port statistics ====================================
Statistics for port 0 ------------------------------
Packets sent:                       10
Packets received:                    9
Packets dropped:                     0
Statistics for port 1 ------------------------------
Packets sent:                        9
Packets received:                   10
Packets dropped:                     0
Aggregate statistics ===============================
Total packets sent:                 19
Total packets received:             19
Total packets dropped:               0
====================================================
^C Signal 2 received, preparing to exit...
Closing port 0... Done
Closing port 1... Done
Bye...
```

And on `benjamin` I can ping the unknown workstation

```
shunghsiyu@benjamin:~> ping -c 5 192.168.3.1
PING 192.168.3.1 (192.168.3.1) 56(84) bytes of data.
64 bytes from 192.168.3.1: icmp_seq=1 ttl=64 time=0.558 ms
64 bytes from 192.168.3.1: icmp_seq=2 ttl=64 time=0.311 ms
64 bytes from 192.168.3.1: icmp_seq=3 ttl=64 time=0.375 ms
64 bytes from 192.168.3.1: icmp_seq=4 ttl=64 time=0.396 ms
64 bytes from 192.168.3.1: icmp_seq=5 ttl=64 time=0.309 ms

--- 192.168.3.1 ping statistics ---
5 packets transmitted, 5 received, 0% packet loss, time 4086ms
rtt min/avg/max/mdev = 0.309/0.389/0.558/0.090 ms
```

Below is the setup I used setup to test l2fwd, it is a bit strange since I doesn't have immediate physical access to the work stations, thus I use what's available.


```
                                         +--------------------------------+
                                         |                                |
                                         |             unknown            |
                                         |                                |
                                         |                                |
  +-------------------------------+      +------------------------+       |
  |                               |      |                        |       |
  |            david              |      |                        |       |
  |                               |      |                  ?     |       |
  |         run `l2fwd`           |      |192.168.3.1             |       |
  |      +--------------+---------+      +---------+              |       |
  |      |  0000:04:00.0|         +------+    ?    |              |       |
  |      |              +---------+      +---------+--------------+-------+
  |      | X520-2                 |                                        
  |      |                        |                                        
  |      |              +---------+      +------------------------+-------+
  |      |  0000:04:00.1|         +------+ixgbe1   |0000:04:00.1  |       |
  |      +--------------+---------+      |---------+              |       |
  +-------------------------------+      |192.168.3.3      X520-2 |       |
                                         |                        |       |
                                         +---------+              |       |
                                         |         |              |       |
                                         +---------+--------------+       |
                                         |                                |
                                         |            benjamin            |
                                         |                                |
                                         |      run `ping 192.168.3.1`    |
                                         +--------------------------------+

```

**Comment by shunghsiyu** [comment]
So it ooks like link callback event is not required for my setup to work. Hmm...
Maybe l2fwd works because it [continuously check the latest link status when waiting in `check_all_ports_link_status()`](https://github.com/DPDK/dpdk/blob/v20.05/examples/l2fwd/main.c#L466), while in TRex we do not do that in [`wait_x_sec()`](https://github.com/cisco-system-traffic-generator/trex-core/blob/v2.83/src/main_dpdk.cpp#L6598)?

**Comment by hhaim** [comment]
@shunghsiyu correct. The same DPDK the same configuration (with ```--software```). The link code is located below ixgbe/base which we didn't change and it does not relate to packet/FID code.

The code is almost of the same (except minor log, ERRO->INFO). 
I have looked into your NIC ID. I have it in the lab and it works fine, maybe it is related to the SFP+ version. I agree that there is an issue as you not the only one reported.

TRex read the link status the same way (every  0.5 sec) so I don't know what is the diff.
try to remove the code of link statsu from l2fwd

**Comment by hhaim** [comment]
@shunghsiyu I mean that we read the link status after trex the ```wait_x_sec()``` finish. This function is to let the link to be stable

**Comment by shunghsiyu** [comment]
Ah now I see that the link status is checked in `is_all_links_are_up()`
Sorry for the misunderstanding and the premature guesses :sweat: 

I'll change the l2fwd code as you suggested and let you know the outcome

**Comment by shunghsiyu** [comment]
For the record I also tried to let TRex wait longer `-w 30` for the link to stabilize since TRex only waits 2 seconds where as l2fwd waits about 9 seconds. But making TRex wait longer for the link to stabilize does not make a different.

**Comment by shunghsiyu** [comment]
Comparing the difference between TRex v2.83 and DPDK v20.05 I found some commits that are in DPDK but not yet in TRex, of which two of them seem to be required for my setup to work
* dpdk/DPDK@efa8c72f1e net/ixgbe: fix link status inconsistencies
* dpdk/DPDK@d68ab7a9f9 net/ixgbe: fix resource leak after thread exits normally

I'll open a PR that adds these commits

**Comment by hhaim** [comment]
@shunghsiyu Understood now. I was mistaken. TRex is based on dpdk 2002 and **not** on 2005 so the comparison was wrong..
Actually, I was looking into the dpdk source code back then and I missed those commit.

**Comment by hhaim** [comment]
will be fixed in v2.84

**Comment by shunghsiyu** [comment]
@hhaim thanks!
